File: ligature-in-memory/src/main/scala/dev/ligature/inmemory/InMemoryQueryTx.scala
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */

package dev.ligature.inmemory

import dev.ligature.*

case class Test2(val x: Int)

/** Represents a QueryTx within the context of a Ligature instance and a single
  * Dataset
  */
// class InMemoryQueryTx(private val store: DatasetStore) extends QueryTx {

//   /** Returns all PersistedTriples in this Dataset. */
//   def allTriples(): Iterator[Triple] = store.edges.iterator

//   /** Returns all PersistedTriples that match the given criteria. If a
//     * parameter is None then it matches all, so passing all Nones is the same as
//     * calling allTriples.
//     */
//   override def matchTriples(
//       entity: Option[LigatureValue.Word],
//       attribute: Option[LigatureValue.Word],
//       value: Option[LigatureValue]
//   ): Iterator[Triple] = {
//     var res = store.edges.iterator
//     if (entity.isDefined) {
//       res = res.filter(_.entity == entity.get)
//     }
//     if (attribute.isDefined) {
//       res = res.filter(_.attribute == attribute.get)
//     }
//     if (value.isDefined) {
//       res = res.filter(_.value == value.get)
//     }
//     res
//   }

//  /** Returns all PersistedTriples that match the given criteria. If a
//    * parameter is None then it matches all.
//    */
//  override def matchTriplesRange(
//      source: Option[Word],
//      label: Option[Word],
//      range: dev.ligature.Range
//  ): Iterator[Triple] = {
//    var res = Stream.emits(store.edges.toSeq)
//    if (source.isDefined) {
//      res = res.filter(_.source == source.get)
//    }
//    if (label.isDefined) {
//      res = res.filter(_.label == label.get)
//    }
//    res = res.filter { ps =>
//      val testValue = ps.target
//      (testValue, range) match {
//        case (String(v), StringRange(start, end)) =>
//          v >= start && v < end
//        case (Int(v), IntRange(start, end)) =>
//          v >= start && v < end
//        case _ => false
//      }
//    }
//    res
//  }
// }

File: ligature-in-memory/src/main/kotlin/dev/ligature/inmemory/InMemoryQueryTx.kt
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */

package dev.ligature.inmemory

import arrow.core.Either
import dev.ligature.*
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.asFlow
import kotlinx.coroutines.flow.filter
import kotlinx.coroutines.flow.map

/** Represents a QueryTx within the context of a Ligature instance and a single Dataset */
class InMemoryQueryTx(private val store: DatasetStore) : QueryTx {
    /** Returns all PersistedStatements in this Dataset. */
    override suspend fun allStatements(): Flow<Either<LigatureError, Statement>> {
        return store.statements.map { Either.Right(it) }.asFlow()
    }

    /** Returns all PersistedStatements that match the given criteria.
     * If a parameter is None then it matches all, so passing all Nones is the same as calling allStatements. */
    override suspend fun matchStatements(
            entity: Entity?,
            attribute: Attribute?,
            value: Value?,
            context: Entity?
    ): Flow<Either<LigatureError, Statement>> {
        var res = store.statements.asFlow()
        if (entity != null) {
            res = res.filter { it.entity == entity }
        }
        if (attribute != null) {
            res = res.filter { it.attribute == attribute }
        }
        if (value != null) {
            res = res.filter { it.value == value }
        }
        if (context != null) {
            res = res.filter { it.context == context }
        }
        return res.map { Either.Right(it) }
    }

    /** Retuns all PersistedStatements that match the given criteria.
     * If a parameter is None then it matches all. */
    override suspend fun matchStatementsRange(
            entity: Entity?,
            attribute: Attribute?,
            range: Range,
            context: Entity?
    ): Flow<Either<LigatureError, Statement>> {
        var res = store.statements.asFlow()
        if (entity != null) {
            res = res.filter { it.entity == entity }
        }
        if (attribute != null) {
            res = res.filter { it.attribute == attribute }
        }
        res = res.filter { s ->
            val value = s.value
            when {
                value is StringLiteral && range is StringLiteralRange   -> value.value >= range.start && value.value < range.end
                value is FloatLiteral && range is FloatLiteralRange     -> value.value >= range.start && value.value < range.end
                value is IntegerLiteral && range is IntegerLiteralRange -> value.value >= range.start && value.value < range.end
                else                                                    -> false
            }
        }
        if (context != null) {
            res = res.filter { it.context == context }
        }
        return res.map { Either.Right(it) }
    }
}
